tests/typecheck/contrib/auth/test_decorators.yml

```diff
@@ -35,6 +35,28 @@
         @user_passes_test(lambda u: u.get_username().startswith('super'))
         def view_func(request: HttpRequest) -> HttpResponse: ...
         reveal_type(view_func)  # N: Revealed type is "def (request: django.http.request.HttpRequest) -> django.http.response.HttpResponse"
+-   case: user_passes_test_custom_user_model
+    main: |
+        from typing import Union
+        from django.contrib.auth.decorators import user_passes_test
+        from django.contrib.auth.models import AnonymousUser
+        from django.http import HttpRequest, HttpResponse
+        from users.models import User
+        def check_user(user: Union[User, AnonymousUser]) -> bool: ...
```

Comment:
What about `def check_user2(user: User) -> bool: ...` ?

Reply:
And `def check_user(user: Union[ AnonymousUser, User]) -> bool: ...`

Reply:
The former would be an error, because it doesn't accept `AnonymousUser`, while the latter is perfectly valid. Note that this change only changes the signature of `user_passes_test`, all the type checking is done by mypy. So basically the signature is changed from:

```python
def user_passes_test(test_func: Callable[[AbstractBaseUser, AnonymousUser], bool]], ...):
```

to:
```python
def user_passes_test(test_func: Callable[[settings.AUTH_USER_MODEL, AnonymousUser], bool]], ...):
```

So anything valid matching the first signature would still match the second one, but the second one allows writing more specific types (using your configured user model).

I added a few more tests cases to demonstrate this.